pubsub/gossipsub/gossipsub-v1.2.md

```diff
@@ -29,32 +49,95 @@ This section lists the configuration parameters that control the behaviour of th
 | Parameter | Description  | Reasonable Default |
 | -------- | -------- | -------- |
 | `D_non_choke`     | The minimum number of peers in a mesh that must remain unchoked. | `D_lo` |
-| `choke_heartbeat_interval` | The number of heartbeats before assessing and applying `CHOKE`/`UNCHOKE` control messages and adding `D_max_add`. | 20 |
-| `D_max_add` | The maximum number of peers to add into the mesh (from fanout) if they are performing well per `choke_heartbeat_interval`. | 1 |
-| `choke_duplicates_threshold` | The minimum number of duplicates as a percentage of received messages a peer must send before being eligible of being `CHOKE`'d. | 60 |
+| `choke_heartbeat_interval` | The number of heartbeats before assessing and applying `CHOKE`/`UNCHOKE` control messages and adding peers to the mesh. | 20 |
 | `choke_churn` | The maximum number of peers that can be `CHOKE`'d or `UNCHOKE`'d in any `choke_heartbeat_interval`. | 2 |
-|` unchoke_threshold` | Determines how aggressively we unchoke peers. The percentage of messages that we receive in the `choke_heartbeat_interval` that were received by gossip from a choked peer. | 50 |
-| `fanout_addition_threshold` | How aggressively we add peers from the fanout into the mesh. The percentage of messages that we receive in the `choke_heartbeat_interval` that were received from a fanout peer. | 10 |
+|` unchoke_churn` | Determines how aggressively we unchoke peers. The number of peers per `choke_heartbeat_interval` that can be unchoked on an individual mesh. | 2 |
+| `mesh_addition_churn` | How aggressively we add peers from into the mesh. The number of peers per `choke_heartbeat_interval` that can be added to an individual mesh. | 1 |
+
+## Implementation Notes
+
+The actual strategy for choking/unchoking peers is left to each implementation
+and potentially application user. Although the strategies for choking/unchoking
+can be generic, a few useful strategies are listed in the appendix for implementers.
+
+## The CHOKE Message
+
+Every `choke_heartbeat_interval` the router applies its choking strategy to a
+set of collected metrics of recent messages, in order to decide if any of its
+mesh peers should be choked. The router should send no more
+than `choke_churn` `CHOKE` messages to peers per mesh topic. The router should
+also ensure that `D_non_choke` peers remain unchoked in each mesh topic.
+
+Upon receiving a `CHOKE` message, the router MUST no longer forward messages to
+the peer that sent the `CHOKE` message, while it is still in the mesh. Instead
+it MUST always send an IHAVE message (provided it does not hit the IHAVE
```

Comment:
pedanticness note: if there were no messages, then it is not required to send anything.